File: s3RubySDK/sample/BucketLoggingSerialTesting.rb
# Set the logging parameters for a bucket and to specify permissions for who can view and modify the logging parameters. 
# To set the logging status of a bucket, you must be the bucket owner.

def bucketloggingserialtesting(s3, buckets, creds)
  
  cleanup
  
  
  resp = s3.create_bucket({
    acl: "log-delivery-write ", # accepts private, public-read, public-read-write, authenticated-read, log-delivery-write
    bucket: buckets[0], # required
    create_bucket_configuration: {
      location_constraint: "ap-northeast-1", # accepts EU, eu-west-1, us-west-1, us-west-2, ap-southeast-1, ap-southeast-2, ap-northeast-1, sa-east-1, cn-north-1, eu-central-1
    }
    })
    
#  resp = s3.get_bucket_acl({
#    bucket: buckets[0], # required
#  })
#
#  puts resp.grants #=> Array
#  puts resp.owner.display_name #=> String
#  puts resp.owner.id #=> String

s3.put_bucket_acl({
  bucket: buckets[0],
  access_control_policy: {
    grants: [
      {
        grantee: {
          type: "Group", # required, accepts CanonicalUser, AmazonCustomerByEmail, Group
          uri: "http://acs.amazonaws.com/groups/s3/LogDelivery",
        },
        permission: "WRITE", # accepts FULL_CONTROL, WRITE, WRITE_ACP, READ, READ_ACP
      },{
        grantee: {
          type: "Group", # required, accepts CanonicalUser, AmazonCustomerByEmail, Group
          uri: "http://acs.amazonaws.com/groups/s3/LogDelivery",
        },
        permission: "READ_ACP", # accepts FULL_CONTROL, WRITE, WRITE_ACP, READ, READ_ACP
        },{
        grantee: {
          type: "CanonicalUser", # required, accepts CanonicalUser, AmazonCustomerByEmail, Group
          id: creds['Owner ID'],
        },
        permission: "FULL_CONTROL", # accepts FULL_CONTROL, WRITE, WRITE_ACP, READ, READ_ACP
      },
    ],
    owner: {
      display_name: creds['Owner Name'],
      id: creds['Owner ID'],
    },
  },
})

    
  resp = s3.put_bucket_logging({
    bucket: buckets[0], # required
    bucket_logging_status: { # required
      logging_enabled: {
        target_bucket: buckets[0],
        target_grants: [
          {
            grantee: {
              id: creds['Grantee Canonical ID'],
              type: "CanonicalUser", # required, accepts CanonicalUser, AmazonCustomerByEmail, Group
            },
            permission: "READ", # accepts FULL_CONTROL, READ, WRITE
          },
        ],
        target_prefix: "logs/",
      },
    },
#    content_md5: "ContentMD5",
  })
 
  
  resp = s3.get_bucket_logging({
    bucket: buckets[0], # required
  })
  puts resp.logging_enabled.target_bucket #=> String
#  puts resp.logging_enabled.target_grants #=> Array
  puts resp.logging_enabled.target_grants[0].grantee.display_name #=> String
#  resp.logging_enabled.target_grants[0].grantee.email_address #=> String
  puts resp.logging_enabled.target_grants[0].grantee.id #=> String
  puts resp.logging_enabled.target_grants[0].grantee.type #=> String, one of "CanonicalUser", "AmazonCustomerByEmail", "Group"
#  resp.logging_enabled.target_grants[0].grantee.uri #=> String
  puts resp.logging_enabled.target_grants[0].permission #=> String, one of "FULL_CONTROL", "READ", "WRITE"
  puts resp.logging_enabled.target_prefix #=> String

  # clean all buckets
  cleanup
end

File: s3JavaScriptSDK/sample/BucketLoggingSerialTesting.js
// 1. set ACL (grant LogDelivery Permission write and WRITE_ACP)
// 2. put bucket Logging
// 3. get bucket Logging

function bucketLoggingTest (){
	return Q.fcall(function (){
		var deferred = Q.defer();
		console.log('-----------------------------------------');
		console.log('bucketLoggingTest begin!');
		deferred.resolve();
		return deferred.promise;
	})
	.then(function(){
		var params = {
				  Bucket: testBucketName, /* required */
				  //ACL: 'public-read', //private | public-read | public-read-write | authenticated-read',
				  AccessControlPolicy: {
				    Grants: [
				      {
				        Grantee: {
				          Type: 'Group', /* required */
				          URI: 'http://acs.amazonaws.com/groups/s3/LogDelivery'
				        },
				        Permission: 'WRITE_ACP',//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
				      },
				      {
				        Grantee: {
				          Type: 'Group', /* required */
				          URI: 'http://acs.amazonaws.com/groups/s3/LogDelivery'
				        },
				        Permission: 'WRITE'//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
				      },
				      {
				        Grantee: {
				          Type: 'Group', /* required */
				          URI: 'http://acs.amazonaws.com/groups/s3/LogDelivery'
				        },
				        Permission: 'READ_ACP'//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
				      },
						{
							Grantee: {
								Type: 'CanonicalUser', /* required */
								ID: ownerID
							},
							Permission: 'FULL_CONTROL',//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
						}
				       // more items 
				    ],
				    Owner: {
				      DisplayName: ownerName,
				      ID: ownerID
				    }
				  },
				};
		return putBucketAcl(params, s3);
	})
	.then(function(){
		var params = {
				  Bucket: testBucketName /* required */
				};
		return getBucketLogging(params, s3);
	})
	.then(function(){
		var params = {
				  Bucket: testBucketName, /* required */
				  BucketLoggingStatus: { /* required */
				    LoggingEnabled: {
				      TargetBucket: testBucketName,
				      TargetGrants: [
				        {
				          Grantee: {
				            Type: 'CanonicalUser', /* required */
				            // DisplayName: 'STRING_VALUE',
				            // EmailAddress: 'STRING_VALUE',
				            ID: granteeId,
				            // URI: ''
				          },
				          Permission: 'READ'//FULL_CONTROL | READ | WRITE
				        },
						  {
							  Grantee: {
								  Type: 'CanonicalUser', /* required */
								  ID: ownerID
							  },
							  Permission: 'FULL_CONTROL',//'FULL_CONTROL | WRITE | WRITE_ACP | READ | READ_ACP'
						  }
				        /* more items */
				      ],
				      TargetPrefix: 'logs/'
				    }
				  },
				  // ContentMD5: 'STRING_VALUE'
				};
		return putBucketLogging(params, s3);
	})
	.then(function(){
		var params = {
				  Bucket: testBucketName /* required */
				};
		return getBucketLogging(params, s3);
	});

}


function putBucketLogging(params, s3){
	var deferred = Q.defer();

	s3.putBucketLogging(params, function(err, data) {
		  if (err) console.log(err, err.stack); // an error occurred
		  else     {
		  	console.log(data);
		  	console.log('put bucket logging succeed');
			deferred.resolve(data);
		  }          
		});
	return deferred.promise;
}

function getBucketLogging(params, s3){
	var deferred = Q.defer();

	s3.getBucketLogging(params, function(err, data) {
		  if (err) console.log(err, err.stack); // an error occurred
		  else     {
		  	console.log(data);
			deferred.resolve(data);
		  }           // successful response
		});
	return deferred.promise;

}
